specs/HubProtocol.md

```diff
@@ -0,0 +1,225 @@
+
+There are two encodings of the SignalR protocol: [JSON](http://www.json.org/) and [Protocol Buffers](https://developers.google.com/protocol-buffers/). Only one format can be used for the duration of a connection, and the format must be negotiated in advance (i.e. using a QueryString value, Header, or other indicator). However, each format shares a similar overall structure.
+
+In the SignalR protocol, the following types of messages can be sent:
+
+* `Invocation` Message - Indicates a request to invoke a particular method (the Target) with provided Arguments on the remote endpoint.
+* `Result` Message - Indicates results from a previous Invocation message.
+* `Completion` Message - Indicates a previous Invocation has completed, and no further Result messages will be received.
+
+In order to perform a single invocation, Caller follows the following basic flow:
+
+1. Allocate a unique `Invocation ID` value (arbitrary string, chosen by the Caller) to represent the invocation
+2. Send an `Invocation` message containing the `Invocation ID`, the name of the `Target` being invoked, and the `Arguments` to provide to the method.
+3. Wait for a `Result` or `Complete` message with a matching `Invocation ID`
+4. If a `Completion` message arrives, go to 8
+5. If the `Result` message has an error, dispatch the error to the application (i.e. throw an `Exception`, etc.) and complete the invocation immediately.
+6. If the `Result` message has a payload, dispatch the payload to the application (i.e. by yielding a result to an `IObservable`, or by collecting the result for dispatching in step 8)
+7. Go to 3
+8. Complete the invocation, yielding that completion to the Caller (i.e. by completing a `Task` or `IObservable`)
+
+The `Target` of an `Invocation` message must refer to a specific method, overloading is **not** permitted. In the .NET Binder, the `Target` value for a method is defined as the simple name of the Method (i.e. without qualifying type name, since a SignalR endpoint is specific to a single Hub class).
+
+**NOTE**: `Invocation ID`s are arbitrarily chosen by the Caller and the Callee is expected to use the same string in all response messages. Callees may establish reasonable limits on `Invocation ID` lengths and terminate the connection when an `Invocation ID` that is too long is received.
+
+## Multiple Results
+
+The SignalR protocol allows for multiple `Result` messages to be transmitted in response to an `Invocation` message, and allows the receiver to dispatch these results as they arrive, to allow for streaming data from one endpoint to another.
+
+On the Callee side, it is up to the Callee's Binder to determine if a method call will yield multiple results. For example, in .NET, a return type of `IEnumerable` or `IObservable` will yield multiple results, but other return types will yield a single result. The Callee's Binder must encode each result in separate `Result` messages, indicating the end of results by sending a `Completion` message.
```

Comment:
There's a lot in here that is up to the Binder, which is the language-specific piece that maps from language concepts to SignalR protocol concepts. For example, we expect the .NET Binder to consider `IEnumerable<T>` returns as a single value (I'll update the spec), partly for this reason. We may end up using an attribute or something to indicate a fully streamed result.

This spec doesn't technically cover the Binder, it just provides examples from the .NET Binder.